scala-xml/src/main/scala/org/http4s/scalaxml/ElemInstances.scala

```diff
@@ -51,19 +55,54 @@ trait ElemInstances {
   )(implicit F: Concurrent[F]): EntityDecoder[F, Stream[F, XmlEvent]] =
     EntityDecoder.decodeBy(MediaType.text.xml, MediaType.text.html, MediaType.application.xml) {
       msg =>
-        DecodeResult.successT(msg.bodyText.through(fs2.data.xml.events(includeComments)))
+        DecodeResult.successT(
+          msg.body.pull
+            .unconsN(64, allowFewer = true)
+            // look at the first bytes to check for BOM or encoding spec in the prolog
+            .flatMap {
+              case Some((chunk, tail)) if chunk.take(2) === `BOM-BE` =>
+                tail
+                  .cons(chunk.drop(2)) // drop BOM
+                  .through(decodeWithCharset(Charset.`UTF-16BE`.nioCharset))
+                  .pull
+                  .echo
+              case Some((chunk, tail)) if chunk.take(2) === `BOM-LE` =>
+                tail
+                  .cons(chunk.drop(2)) // drop BOM
+                  .through(decodeWithCharset(Charset.`UTF-16LE`.nioCharset))
+                  .pull
+                  .echo
+              case Some((chunk, tail)) if msg.charset.isEmpty =>
+                // guard: xml prolog is only authoritative if content type is missing, see RFC 7303 §8.8
+                guessCharset(new String(chunk.toArray, `UTF-8`.nioCharset))
+                  .fold(
+                    Pull.raiseError[F](_),
+                    charset =>
+                      tail.cons(chunk).through(decodeWithCharset(charset.nioCharset)).pull.echo,
+                  )
+              case Some((chunk, tail)) =>
+                tail
+                  .cons(chunk)
+                  .through(decodeWithCharset(msg.charset.getOrElse(`UTF-8`).nioCharset))
+                  .pull
+                  .echo
+              case None => Pull.done
```

Comment:
Sorry if I missed it, I don't think we have test coverage for the BOM cases? IDK how important they are but if the RFC has any examples would be good to add.

Reply:
:thumbsup: There's a test failing if the BOM-handling code is missing, but I added two explicit tests (one per endianness) for BOMs.

Reply:
:thumbsup: There's a test failing if the BOM-handling code is missing, but I added two explicit tests (one per endianness) for BOMs.

Reply:
Awesome thanks!